W-14927671  Pdk ga branch gr (#464)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-configure-features-libraries.adoc

= Using Third-Party Libraries
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

NOTE: To view an example policy project that uses third-party libraries, see https://github.com/mulesoft/pdk-custom-policy-examples/blob/main/crypto/README.md[Crypto Policy Example^].

Proxy Wasm defines a low level binary application interface (abi), that limits the system calls a library can make. All third-party libraries included in your policy must be compatible with the `wasm32-wasi` Rust compilation target.

To define a library such as `serde_urlencoded`, add it to the dependency list in `cargo.toml`:

[source,toml]
----
[dependencies]
...
serde_urlencoded = "0.7.0"
----

Defining libraries enables you to use functions in `lib.rs` such as:

[source,Rust]
----
serde_urlencoded::to_string([("token", "myToken")])
----

Libraries that interact with an external service like a database, or perform a system call like reading a file, are usually not compatible with the `wasm32-wasi` target. 

> If your policy interacts with external services, use the Flex Gateway exposed HttpClient. For more information about performing an HTTP call, see xref:policies-pdk-configure-features-http-request.adoc[].

Reviewer: ```suggestion
If your policy interacts with external services, use the Flex Gateway exposed `HttpClient`. For more information about performing an HTTP call, see xref:policies-pdk-configure-features-http-request.adoc[].
```

> It is possible that some libraries compile properly to the `wasm32-wasi` target but don't work properly when deployed to Flex Gateway. Example errors of this are:

Reviewer: ```suggestion
It is possible that some libraries compile properly to the `wasm32-wasi` target but don't work properly when deployed to Flex Gateway. Example errors include:
```

* `Failed to load Wasm module due to a missing import: ...`
* `Wasm VM failed to initialize Wasm code`
* `Plugin configured to fail closed failed to load`

> In this case, contact the owners of the third-party library or use a different library.

Reviewer: ```suggestion
In these cases, contact the owners of the third-party library, or use a different library.
```
